crates/nu-cli/src/shell/shell.rs

```diff
@@ -6,13 +6,15 @@ use crate::commands::ls::LsArgs;
 use crate::commands::mkdir::MkdirArgs;
 use crate::commands::move_::mv::Arguments as MvArgs;
 use crate::commands::rm::RemoveArgs;
+use crate::completion;
 use crate::prelude::*;
 use crate::stream::OutputStream;
+
 use encoding_rs::Encoding;
 use nu_errors::ShellError;
 use std::path::PathBuf;
 
-pub trait Shell: std::fmt::Debug {
+pub trait Shell: completion::Completer + std::fmt::Debug {
```

Comment:
I understand why we had a shell do completion, but I think we can eventually move ourselves away from this.

I'm thinking a `Completer` can have the current shell as part of the context. If it needs to complete a path, it can ask the shell to `ls`, and filter as necessary. If we're concerned about performance, we could consider an arg for `ls` to filter things down.

Reply:
Chatting a little with @jonathandturner, perhaps it makes sense to continue making `Shell` also be a `Completer`, but it would only be able to complete a file/path shape. We may end up having a mapping of shape to completer somewhere else.